### PR TITLE
Expose container start time in kubelet /metrics/resource endpoint

### PR DESCRIPTION
**Pull request by sanwishe**
Signed-off-by: sanwishe <jiang.mingzhi35@zte.com.cn>

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Expose container start time in kubelet /metrics/resource endpoint
#### Which issue(s) this PR fixes:

Fixes #
#101851
#### Does this PR introduce a user-facing change?

```release-note
Expose container start time as container_start_time_seconds in kubelet /metrics/resource endpoint
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
"NONE"
```

### CODE REVIEW

File: pkg/kubelet/metrics/collectors/resource_metrics_test.go
Comment (wzshiming) [inline pkg/kubelet/metrics/collectors/resource_metrics_test.go:179]: I think if it is a floating point number, the precision will be lost. Can you add a case to verify this?

**Review by wzshiming — CHANGES_REQUESTED** [review]

File: pkg/kubelet/metrics/collectors/resource_metrics.go
Comment (serathius) [inline pkg/kubelet/metrics/collectors/resource_metrics.go:167]: I think we should either use same format as:
* process_start_time metric or
* [timestampedMetric](https://github.com/kubernetes/kubernetes/blob/14e4ed07cf81a8a273c63943c29f9f54761daebb/vendor/github.com/prometheus/client_golang/prometheus/metric.go#L157)

Can you compare those two formats so we can discuss it in this PR? Would be great to see some real examples

File: pkg/kubelet/metrics/collectors/resource_metrics_test.go
Comment (serathius) [inline pkg/kubelet/metrics/collectors/resource_metrics_test.go:114]: Can you set here some real data so it's easier to verify it the result makes sense?

File: pkg/kubelet/metrics/collectors/resource_metrics.go
Comment (sanwishe) [reply pkg/kubelet/metrics/collectors/resource_metrics.go:167]: hi, @serathius, 

- `process_start_time_seconds` of kubelet does not reported as `timestampedMetrics`,:

```
[root@server-78 ~]# curl -s  http://127.0.0.1:8001/api/v1/nodes/10.235.30.78/proxy/metrics |grep process_start_time
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.62250734682e+09
```

- The `statsSummary` reported by `stats.SummaryProvider` provide no sampling timestamp of `container_start_time` metrics, since `container_start_time` is independent of sampling timestamp, does `time.Now()` make scence?

here is the real example metrics that using `time.Now()` as the sampling timestamp of `container_start_time`:
```
[root@server-78 ~]# curl -s  http://127.0.0.1:8001/api/v1/nodes/10.235.30.78/proxy/metrics/resource |grep  container_start_time
# HELP container_start_time [ALPHA] Start time of container in UNIX time
# TYPE container_start_time gauge
container_start_time{container="calico-kube-controllers",namespace="kube-system",pod="calico-kube-controllers-fcb4577b8-vtmhn"} 1.6225073641000604e+18 1622509662905
container_start_time{container="calico-node",namespace="kube-system",pod="calico-node-fgztf"} 1.6225073746780928e+18 1622509662905
container_start_time{container="kube-apiserver",namespace="kube-system",pod="kube-apiserver-10.235.30.78"} 1.622507351582303e+18 1622509662905
container_start_time{container="kube-controller-manager",namespace="kube-system",pod="kube-controller-manager-10.235.30.78"} 1.6225073518739638e+18 1622509662905
container_start_time{container="kube-proxy",namespace="kube-system",pod="kube-proxy-10.235.30.78"} 1.6225073517178916e+18 1622509662905
container_start_time{container="kube-scheduler",namespace="kube-system",pod="kube-scheduler-10.235.30.78"} 1.6225073515772767e+18 1622509662906
container_start_time{container="test-pvc-busybox",namespace="default",pod="a-test-pvc-sts-0"} 1.6225073717213302e+18 1622509662906
container_start_time{container="test-pvc-busybox",namespace="default",pod="a-test-pvc-sts-1"} 1.6225073684144394e+18 1622509662906
```

If we do not use `timestampedMetrics`, `container_start_time` metrics will be exposed like this:
```
[root@server-78 ~]# curl -s  http://127.0.0.1:8001/api/v1/nodes/10.235.30.78/proxy/metrics/resource |grep  container_start_time
# HELP container_start_time [ALPHA] Start time of container in UNIX time
# TYPE container_start_time gauge
container_start_time{container="calico-kube-controllers",namespace="kube-system",pod="calico-kube-controllers-fcb4577b8-vtmhn"} 1.6225073641000604e+18
container_start_time{container="calico-node",namespace="kube-system",pod="calico-node-fgztf"} 1.6225073746780928e+18
container_start_time{container="kube-apiserver",namespace="kube-system",pod="kube-apiserver-10.235.30.78"} 1.622507351582303e+18
container_start_time{container="kube-controller-manager",namespace="kube-system",pod="kube-controller-manager-10.235.30.78"} 1.6225073518739638e+18
container_start_time{container="kube-proxy",namespace="kube-system",pod="kube-proxy-10.235.30.78"} 1.6225073517178916e+18
container_start_time{container="kube-scheduler",namespace="kube-system",pod="kube-scheduler-10.235.30.78"} 1.6225073515772767e+18
container_start_time{container="test-pvc-busybox",namespace="default",pod="a-test-pvc-sts-0"} 1.6225073717213302e+18
container_start_time{container="test-pvc-busybox",namespace="default",pod="a-test-pvc-sts-1"} 1.6225073684144394e+18
```

File: pkg/kubelet/metrics/collectors/resource_metrics_test.go
Comment (sanwishe) [reply pkg/kubelet/metrics/collectors/resource_metrics_test.go:114]: ok, i will appliy your suggestion.

File: pkg/kubelet/metrics/collectors/resource_metrics.go
Comment (serathius) [reply pkg/kubelet/metrics/collectors/resource_metrics.go:167]: Ok, let's use the same format as process_start_time so ` time since unix epoch in seconds`, but let's make sure to keep the milisecond resolution. This differs a big from the last result you have shown which is in nanoseconds.

Difference is in unit `seconds` vs `nanoseconds` and can be seen in exponential `e+09` vs `e+18`. Before passing time value let's make sure we divide it to 10^9 to change the unit.

Example in seconds
```
process_start_time_seconds 1.62250734682e+09
```

Example In nanoseconds:
```
container_start_time{container="calico-kube-controllers",namespace="kube-system",pod="calico-kube-controllers-fcb4577b8-vtmhn"} 1.6225073641000604e+18
```

Let's also update the description of the metric to reflect the unit.
```
# HELP container_start_time [ALPHA] Start time of container in UNIX time
```
To 
```
# HELP container_start_time [ALPHA] Start time of the container since unix epoch in seconds.
```

Comment (sanwishe) [reply pkg/kubelet/metrics/collectors/resource_metrics.go]: Thanks for your advise, and now the metrics `container_start_time` exposed as the example blow. 
```
# HELP container_start_time [ALPHA] Start time of the container since unix epoch in seconds
# TYPE container_start_time gauge
container_start_time{container="container_a",namespace="namespace_a",pod="pod_a"} 1.624396248302e+09
container_start_time{container="container_a",namespace="namespace_b",pod="pod_b"} 1.624395678302e+09
container_start_time{container="container_b",namespace="namespace_a",pod="pod_a"} 1.624396158302e+09
```

There ara hesitation about the metrics name, should we add the unit `second` to the metrics name, that is `container_start_time_seconds`, so it is more prometheus style?

File: pkg/kubelet/metrics/collectors/resource_metrics_test.go
Comment (MHBauer) [inline pkg/kubelet/metrics/collectors/resource_metrics_test.go:48]: What is Nona?

File: pkg/kubelet/metrics/collectors/resource_metrics.go
Comment (MHBauer) [inline pkg/kubelet/metrics/collectors/resource_metrics.go:167]: This construction seems a bit weird to me. 
We divide billionths by millionths, convert and then divide by thousandths.
Is there a reason for the multiple steps? We can see up a few lines that StartTime has a Unix() for the second, why are we calculating it ?

Comment (MHBauer) [reply pkg/kubelet/metrics/collectors/resource_metrics.go:167]: I read, https://github.com/kubernetes/kubernetes/pull/102444#discussion_r654928012 but am still confused.

**Review by MHBauer — COMMENTED** [review]
good luck. Thanks for the PR.

File: pkg/kubelet/metrics/collectors/resource_metrics.go
Comment (sanwishe) [reply pkg/kubelet/metrics/collectors/resource_metrics.go]: Sorry, but metrics `container_start_time` is requested to second as unit but keep the millisecond resolution. and does this make sense?
```
containerStartTimeSec := float64(s.StartTime.UnixNano()/int64(time.Millisecond)) / 1e3
```

File: pkg/kubelet/metrics/collectors/resource_metrics_test.go
Comment (sanwishe) [reply pkg/kubelet/metrics/collectors/resource_metrics_test.go]: Sorry, it is a misstake, I mean to use nanoseconds, updated

File: pkg/kubelet/metrics/collectors/resource_metrics.go
Comment (serathius) [reply pkg/kubelet/metrics/collectors/resource_metrics.go]: Why not?
```
containerStartTimeSec := float64(s.StartTime.UnixNano()/int64(time.Second))
```

File: pkg/kubelet/metrics/collectors/resource_metrics.go
Comment (serathius) [reply pkg/kubelet/metrics/collectors/resource_metrics.go]: I think so, it also matches `process_start_time_seconds`. Good catch.

File: pkg/kubelet/metrics/collectors/resource_metrics.go
Comment (sanwishe) [reply pkg/kubelet/metrics/collectors/resource_metrics.go]: This will lose the millisecond resolution result as `1.624396158e+09`, but required as format `1.624396158302e+09`.

Comment (serathius) [reply pkg/kubelet/metrics/collectors/resource_metrics.go]: Right, sorry I forgot we want to have milisecond resolution

Have you tried converting to float first?
```
float64(s.StartTime.UnixNano()) / float64(time.Second)
```

Comment (sanwishe) [reply pkg/kubelet/metrics/collectors/resource_metrics.go]: This works absolutely, and result as `1.6243962783020916e+09` which has precise to microsecond. If we do not mind it, this expression looks more simple and clear.

**Comment by serathius** [comment]
@sanwishe Please remember that code freeze is closing in (July 8, 2021). Could you try to merge it in?

**Comment by sanwishe** [comment]
> @sanwishe Please remember that code freeze is closing in (July 8, 2021). Could you try to merge it in before?

Sorry, I'm looking for approvers to review and merge this.

File: pkg/kubelet/metrics/collectors/resource_metrics.go
Comment (serathius) [reply pkg/kubelet/metrics/collectors/resource_metrics.go]: LG

**Comment by sanwishe** [comment]
@dashpole @klueska, Would you please approve this pr if no updates needed? thank you very much.

**Comment by serathius** [comment]
ping @klueska

**Comment by serathius** [comment]
@kubernetes/sig-node-pr-reviews

**Review by mrunalp — APPROVED** [review]

**Comment by harche** [comment]
@sanwishe @serathius @SergeyKanzhelev https://github.com/kubernetes/kubernetes/issues/103428

File: pkg/kubelet/metrics/collectors/resource_metrics.go
Comment (fuweid) [inline pkg/kubelet/metrics/collectors/resource_metrics.go:167]: just use `float64(s.StartTime.Unix())` here?

**Pull request merged**
